pvlib/modelchain.py

```diff
@@ -877,19 +883,46 @@ def infer_spectral_model(self):
         params = _common_keys(module_parameters)
         if {'A4', 'A3', 'A2', 'A1', 'A0'} <= params:
             return self.sapm_spectral_loss
-        elif ((('Technology' in params or
-                'Material' in params) and
-               (self.system._infer_cell_type() is not None)) or
-              'first_solar_spectral_coefficients' in params):
+        elif "first_solar_spectral_coefficients" in params:
+            # user explicitly sets spectral coefficients
             return self.first_solar_spectral_loss
+        elif (
+            # cell type is known or can be inferred
+            ("Technology" in params or "Material" in params)
+            and (self.system._infer_cell_type() is not None)
+        ):
+            # This suggests models that provide default parameters per cell
+            # type can be used. However, some of them depend on other weather
+            # parameters, so we need to check if they are available.
+            # At this point, weather may not be available, so let's take that
+            # into account.
+            if self.results is not None and self.results.weather is not None:
```

Comment:
why check `ModelChain.results` instead of passing `weather` to `infer_spectral_model`?

Reply:
You've got a point